docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,1063 @@
+teams and CI pipelines where the configuration needs to be stored and synced
+externally.
+
+With the management cluster operator, we aim to address these use cases by
+introducing an operator that handles the lifecycle of providers within the
+management cluster based on a declarative API.
+
+### Goals
+
+- Define an API that enables declarative management of the lifecycle of
+  Cluster API and all of its providers.
+- Support air-gapped environments through sufficient documentation initially.
+- Identify and document differences between clusterctl CLI and the operator in
+  managing the lifecycle of providers, if any.
+- Define how the clusterctl CLI should be changed in order to interact with
+  the management cluster operator in a transparent and effective way.
+- To support the ability to upgrade from a v1alpha3 based version (v0.3.[TBD])
+  of Cluster API to one managed by the operator.
+
+### Non-Goals/Future Work
+
+- Implement an operator driven version of `clusterctl move`.
+- Manage cert-manager using the operator.
+- Support multiple installations of the same provider within a management
+  cluster in light of [issue 3042] and [issue 3354].
+- Support any template processing engines.
+- Support the installation of v1alpha3 providers using the operator.
+
+## Proposal
+
+### User Stories
+
+1. As an admin, I want to use a declarative style API to operate the Cluster
+   API providers in a management cluster.
+1. As an admin, I would like to have an easy and declarative way to change
+   controller settings (e.g. enabling pprof for debugging).
+1. As an admin, I would like to have an easy and declarative way to change the
+   resource requirements (e.g. such as limits and requests for a provider
+   deployment).
+1. As an admin, I would like to have the option to use clusterctl CLI as of
+   today, without being concerned about the operator.
+1. As an admin, I would like to be able to install the operator using kubectl
+   apply, without being forced to use clusterctl.
+
+### Implementation Details/Notes/Constraints
+
+#### Existing API Types Changes
+
+The existing `Provider` type used by the clusterctl CLI will be deprecated and
+its instances will be migrated to instances of the new API types as defined in
+the next section.
+
+The management cluster operator will be responsible for migrating the existing
+provider types to support GitOps workflows excluding `clusterctl`.
+
+#### New API Types
+
+These are the new API types being defined.
+
+There are separate types for each provider type - Core, Bootstrap,
+ControlPlane, and Infrastructure. However, since each type is similar, their
+Spec and Status uses the shared types - `ProviderSpec`,  `ProviderStatus`
+respectively.
+
+If you prefer to see how the API can be used instead of reading the type
+definition feel free to jump to the [Example API Usage
+section](#example-api-usage)
+
+```golang
+// CoreProvider is the Schema for the CoreProviders API
+type CoreProvider struct {
```

Comment:
So we actually did define separate Specs for each provider type but since they were using all using the `ProviderSpec`, the suggestion was to remove the specific `CoreProviderSpec` and just have it as `ProviderSpec`. In the future, if we need to include specific properties we can embed the `ProviderSpec` type and this will not break compatibility.
```golang
type CoreProviderSpec struct {
  ProviderSpec
  AnotherField string
}
```

For the reasoning for having separate types like `CoreProvider`, `BootstrapProvider` etc. this is because we still need a way to identify the providers. Doing something like below doesn't seem to follow k8s patterns _at least IMO_.
```golang
type Provider struct {
  Type string
}
```

Reply:
sgtm